### DISCUSSION TITLE
DecisionTables with `collect` hit policy overwite outputs

### DISCUSSION THREAD

**Discussion by edward-morgan** [gorules/zen#268] [Q&A] [answered]
Hi, and thanks for all your hard work on this project! I wanted to bring up a potential bug that our team has run into around Decision Tables with the `collect` hit policy.

Given a decision model like the following: [edge-ordering.json](https://github.com/user-attachments/files/17497506/edge-ordering.json)

where `decisionTable1` has a `collect` hit policy, the order that `decisionTable1` and `expression1` are connected to `response` dictates what is produced.

If `decisionTable1` is connected first:

<img width="1807" alt="image" src="https://github.com/user-attachments/assets/08f7d7ce-35f0-4fa4-93ff-ce8ce08677d6">

If `expression1` is connected first:

<img width="1808" alt="image" src="https://github.com/user-attachments/assets/3529e927-4da5-457f-be16-5e7118761fd4">

This appears to be due to the [overwriting strategy](https://github.com/gorules/zen/blob/%40gorules/zen-engine%400.32.0/core/expression/src/variable/mod.rs#L320) in `merge_variables()`, specifically if merging an array with an object. This problem seems somewhat similar to https://github.com/gorules/zen/discussions/90, but I wanted to bring it up separately since connecting nodes in a certain order will change what data are produced even though everything looks the same in the UI (and the internal graphs are structurally identical).

From reading the earlier discussion, I'm guessing the current strategy would be to fix this in the JDM with a Function node to coalesce the array:

```js
/** @type {Handler} **/
export const handler = async (input) => {
  return {"collected": input};
};
```

Can this be fixed in the engine instead? Alternatively, I can put up a PR updating the README with a note mentioning this and the recommended workaround.

**Comment by ivanmiletic** [comment, marked as answer]
Hi, we are in process of releasing a new features that will solve your issue.

It will allow you to set output of nodes to some field or "path". Also it will allow us to do loops in certain nodes as well as passthrough data merge.
![Screenshot 2024-10-21 at 11 30 47](https://github.com/user-attachments/assets/b9438a7b-507c-4b04-98bc-17049db7814b)

**Comment by stefan-gorules** [comment]
Hi, as this has been released I will move this to a discussion.

**Reply by jakeberg** [reply]
Was this fix added in 0.32.0 for the Node package?

**Reply by ivanmiletic** [reply]
Yes, you can scroll at the end of this page to check: https://docs.gorules.io/docs/loops#output-path

**Comment by edward-morgan** [comment]
That's great, thanks so much!

**Reply by stefan-gorules** [reply]
You are welcome! Forgot to mention here are the docs for your use case:
https://gorules.io/docs/user-manual/decision-modeling/data-flow
https://gorules.io/docs/user-manual/decision-modeling/loops

Most important properties will be `outputPath` and/or `inputPath` in your case.
